### PR TITLE
fixed bytes operations in warplib

### PR DESCRIPTION
**Pull request by AlejandroLabourdette**

### CODE REVIEW

File: warplib/src/warp_memory/bytes.cairo [hunk 1]
```diff
@@ -1,6 +1,62 @@
+use integer::u128_to_felt252;
+use integer::u256_from_felt252;
+use warplib::integer::u256_from_felts;
+use warplib::integer::u256_into_felt_unsafe;
 use warplib::warp_memory::WarpMemory;
 use warplib::warp_memory::WarpMemoryTrait;
 
 trait WarpMemoryBytesTrait{
     fn bytes_to_fixed_bytes32(ref self: WarpMemory, location: felt252) -> u256;
+    fn wm_bytes_to_fixed_bytesN(self: WarpMemory, bytesLoc: felt252, width: felt252) -> felt252 ;
```
Comment (rodrodros) [inline warplib/src/warp_memory/bytes.cairo]: We can now remove the `wm_` prefix, we no longer need it

File: warplib/src/warp_memory/bytes.cairo [hunk 2]
```diff
@@ -1,6 +1,62 @@
...
 }
+
+impl WarpMemoryBytesImpl of WarpMemoryBytesTrait{
+    fn bytes_to_fixed_bytes32(self: WarpMemory, bytesLoc: felt252) -> u256 {
+        // TODO Is this self call to read_felt enough, don't we need to import from WarpMemoryReadTrait??
+        let data_len = self.read_u256(bytesLoc);
```
Comment (rodrodros) [inline warplib/src/warp_memory/bytes.cairo]: length from bytes is now stored as a felt instead of a u256. You should read a single felt from here

File: warplib/src/warp_memory/bytes.cairo [hunk 3]
```diff
@@ -1,6 +1,62 @@
...
 use warplib::warp_memory::WarpMemory;
 use warplib::warp_memory::WarpMemoryTrait;
 
 trait WarpMemoryBytesTrait{
     fn bytes_to_fixed_bytes32(ref self: WarpMemory, location: felt252) -> u256;
+    fn wm_bytes_to_fixed_bytesN(self: WarpMemory, bytesLoc: felt252, width: felt252) -> felt252 ;
 }
+
+impl WarpMemoryBytesImpl of WarpMemoryBytesTrait{
+    fn bytes_to_fixed_bytes32(self: WarpMemory, bytesLoc: felt252) -> u256 {
```
Comment (rodrodros) [inline warplib/src/warp_memory/bytes.cairo]: camelCase to snake_case

File: warplib/src/warp_memory/bytes.cairo [hunk 4]
```diff
@@ -1,6 +1,62 @@
...
+            }
+        } else {
+            let amount_to_read = u128_to_felt252(data_len.low);
+            let low_felt = self._bytes_to_fixed_recursive(bytesLoc + 2, 16, amount_to_read, 0);
+            return u256_from_felts(low_felt, 0);
+        }
+    }
+
+    // TODO This function will convert to fixed but store it in a felt, so will only be saved the last 31 bytes. Is this the approach we want??
+    fn wm_bytes_to_fixed_bytesN(self: WarpMemory, bytesLoc: felt252, width: felt252) -> felt252 {
```
Comment (rodrodros) [inline warplib/src/warp_memory/bytes.cairo]: camelCase to snake_case of `bytesLoc`

File: warplib/src/warp_memory/bytes.cairo [hunk 5]
```diff
@@ -1,6 +1,62 @@
...
+
+    fn _bytes_to_fixed_recursive(self: WarpMemory, bytesDataLoc: felt252, targetWidth: felt252, data_len: felt252, representation: felt252) -> felt252 {
+        if targetWidth == 0 {
+            return representation;
+        }
+        if data_len == 0 {
+            return self._bytes_to_fixed_recursive(bytesDataLoc + 1, targetWidth - 1, data_len, 256 * representation);
+        } else {
+            // TODO Is this self call to read_felt enough, don't we need to import from WarpMemoryReadTrait??
+            let byte = self.read_felt(bytesDataLoc);
```
Comment (rodrodros) [inline warplib/src/warp_memory/bytes.cairo]: self.read_felt to self.read

File: warplib/src/warp_memory/read.cairo [hunk 6]
```diff
@@ -3,10 +3,17 @@ use warplib::warp_memory::WarpMemoryTrait;
...
 impl WarpMemoryReadImpl of WarpMemoryReadTrait {
     fn read_felt(ref self: WarpMemory, index: felt252) -> felt252 {
         self.read(index)
     }
+
+    fn read_u256(ref self: WarpMemory, index: felt252) -> u256 {
+        let low = self.read(index);
+        let high = self.read(index + 1);
+        u256{low, high}
+    }
```
Comment (piwonskp) [inline warplib/src/warp_memory/read.cairo]: We should use MemoryAccessor's `retrieve` instead

File: warplib/src/warp_memory/bytes.cairo (hunk 5 above)
Comment (AlejandroLabourdette) [reply warplib/src/warp_memory/bytes.cairo]: Then why would be needed WarpMemoryReadTrait?

https://github.com/NethermindEth/warp/blob/cac9aa92e682b537b0d02426c8c93594889a3639/warplib/src/warp_memory/read.cairo#L4-L6

Comment (piwonskp) [reply warplib/src/warp_memory/bytes.cairo]: See https://github.com/NethermindEth/warp/pull/1038#discussion_r1181965536

**Comment by piwonskp** [comment]
Let's add some unit testing as I've suggested in https://github.com/NethermindEth/warp/pull/1038#issuecomment-1534522079

File: warplib/src/warp_memory/bytes.cairo [hunk 7]
```diff
@@ -1,6 +1,51 @@
...
+        let u256_literal_32 = u256{low: 32, high: 0};
+        if data_len_u256 >= u256_literal_16 {
+            let low_felt = _bytes_to_fixed_recursive(ref self, location + 1, 16, 16, 0);
+            if data_len_u256 >= u256_literal_32 {
+                let high_felt = _bytes_to_fixed_recursive(ref self, location + 17, 16, 16, 0);
+                return u256_from_felts(low_felt, high_felt); 
+            } else {
+                let high_felt = _bytes_to_fixed_recursive(ref self, location + 17, 16, data_len - 16, 0);
+                return u256_from_felts(low_felt, high_felt); 
+            }
```
Comment (piwonskp) [inline warplib/src/warp_memory/bytes.cairo]: Let's define just data len conditionally since every other argument is the same

File: warplib/src/warp_memory/bytes.cairo [hunk 8]
```diff
@@ -1,6 +1,51 @@
...
+fn _bytes_to_fixed_recursive(ref memory: WarpMemory, bytes_data_loc: felt252, target_width: felt252, data_len: felt252, representation: felt252) -> felt252 {
+    if target_width == 0 {
+        return representation;
+    }
+    if data_len == 0 {
+        return _bytes_to_fixed_recursive(ref memory, bytes_data_loc + 1, target_width - 1, data_len, 256 * representation);
+    } else {
+        let byte = memory.read(bytes_data_loc);
+        return _bytes_to_fixed_recursive(ref memory, bytes_data_loc + 1, target_width - 1, data_len - 1, 256 * representation + byte);
+    }
```
Comment (piwonskp) [inline warplib/src/warp_memory/bytes.cairo]: Imo it'd be more idiomatic to use loop. @rodrigo-pino WYT?

**Comment by piwonskp** [comment]
The PR include commits from @rodrigo-pino branch #1038

**Comment by AlejandroLabourdette** [comment]
@rodrigo-pino could you please take a look at the merge, being more specific: to the warp-memory-test.cairo changes. I give priority to incoming changes in sections that were not addressed by this pr, but seems logic to include some of the former approach.

File: warplib/src/test/fixed_bytes_test.cairo [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    let and_r: bytes32 = fixed_byte32_1.and_bitwise(fixed_byte32_2);
+    let expected_value_and = u256 { low: 171927833395647130874784230521578192964_u128, high: 171927833395647130874784230521578192964_u128 };
+    assert(and_r.value == expected_value_and, 'Unexpected bytes32 bitwise and');
+    let or_r: bytes32 = fixed_byte32_1.or_bitwise(fixed_byte32_2);
+    let expected_value_or = u256 { low: 244380555661758051805493511699583001183_u128, high: 244380555661758051805493511699583001183_u128 };
+    assert(or_r.value == expected_value_or, 'Unexpected bytes32 bitwise or');
+    let xor_r: bytes32 = fixed_byte32_1.exclusive_or_bitwise(fixed_byte32_2);
+    let expected_value_xor = u256 { low: 72452722266110920930709281178004808219_u128, high: 72452722266110920930709281178004808219_u128 };
+    assert(xor_r.value == expected_value_xor, 'Unexpected bytes32 bitwise xor');
+}
```
Comment (piwonskp) [inline warplib/src/test/fixed_bytes_test.cairo]:
```suggestion
}

```

File: warplib/src/test/warp_memory_test.cairo [hunk 10]
```diff
@@ -307,3 +312,206 @@ fn test_store_retrieve() {
...
+    let fixed_byte13: bytes13 = warp_memory.bytes_to_fixed_bytes13(array_pointer);
+    let expected_value = 0x8359dba499164cbe2fc9c86dc7_u128;
+    assert(fixed_byte13.value == expected_value, 'Unexpected bytes13 read');
+
+    let byte0 = fixed_byte13.atIndex(0_u8);
+    assert(byte0.value == 131_u8, 'Unexpected byte 0');
+
+    let byte12 = fixed_byte13.atIndex(12_u8);
+    assert(byte12.value == 199_u8, 'Unexpected byte 12');
+}
```
Comment (piwonskp) [inline warplib/src/test/warp_memory_test.cairo]:
```suggestion
}

```

**Comment by temyurchenko** [comment]
@AlejandroLabourdette could you please rebase on cairo-1.0, it seems that many commits are from the already merged branch didn't disappear

**Comment by temyurchenko** [comment]
Unfortunately, merges only add more commits( You'll have to rebase

File: src/warplib/implementations/types/bytes.ts [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+      fn greater_comparison(self: T, other: T) -> bool; // a > b
+      // Bitwise
+      // TODO: Bitwise is not natively supported for all uN types. We can convert them into the ones that
+      // implement the operation and go back to the original type 
+      fn and_bitwise(self: T, other: T) -> T; // a & b
+      fn or_bitwise(self: T, other: T) -> T; // a | b
+      fn exclusive_or_bitwise(self: T, other: T) -> T; // a ^ b
+      // Bitwise negation is not currently supported
+      // Shift operators are not currently supported
+    }
```
Comment (piwonskp) [inline src/warplib/implementations/types/bytes.ts:28]: Can we store the trait in warplib statically?

Comment (AlejandroLabourdette) [reply src/warplib/implementations/types/bytes.ts:28]: to do so we would need to create a separate file just for that, then we would have 2 files every time we would want to check something about bytes. Also seems better to have trait definition next to its implementations due that this will not be a commonly extended trait

File: .starknet_accounts_devnet/starknet_open_zeppelin_accounts.json
Comment (piwonskp) [inline .starknet_accounts_devnet/starknet_open_zeppelin_accounts.json:1]: Please revert the changes to the file

File: src/warplib/implementations/types/bytes.ts [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+              ? `bytes${length}{ value: self.value | other.value }`
+              : endent`
+              let self_value_felt252 = u${max_width}_to_felt252(self.value);
+              let other_value_felt252 = u${max_width}_to_felt252(other.value);
+              let self_value_u128 = u128_try_from_felt252(self_value_felt252).unwrap();
+              let other_value_u128 = u128_try_from_felt252(other_value_felt252).unwrap();
+              let result_u128 = self_value_u128 | other_value_u128;
+              let result_felt252 = u128_to_felt252(result_u128);
+              bytes${length}{ value: u${max_width}_try_from_felt252(result_felt252).unwrap() }
+            `
```
Comment (cicr99) [inline src/warplib/implementations/types/bytes.ts]: This code is the same for add, or and xor right? The only difference is the operator. I think you can write this body inside of a function that also receives the symbol of the operator and returns the string to be written in the cairo output

**Review by cicr99 — APPROVED** [review]

**Review by piwonskp — APPROVED** [review]
